### DISCUSSION TITLE
Check only visible text

### DISCUSSION THREAD

**Discussion by walseb** [minad/jinx#144] [Ideas]
Currently, commands like `jinx-correct-all`, and `jinx-correct-nearest` tries to correct words that aren't visible (for instance those folded with folding tools, or outline-mode).

While thorough, sometimes, this isn't useful. For instance today, I folded headings containing massive amounts of text in an attempt to exclude them from being spell checked when checking other text.

I could have achieved the same thing by narrowing a bunch of times and spell checking those individual buffers, but that would have taken a long time and be error prone. I ended up just patching Jinx to only check visible text.

Would you be open to a patch that adds a variable `jinx-check-only-visible` to change this behavior?

**Comment by minad** [comment]
Yes, please submit the patch. Jinx usually only updates unfolded regions lazily, but `jinx-correct-all` and `jinx-correct-nearest` enforce spell checking in the whole region, and does not take folded regions into account. We should also change that. Do you have the FSF papers signed, see https://github.com/minad/jinx?tab=readme-ov-file#contributions?

**Reply by minad** [reply]
Ping @walseb?

**Reply by walseb** [reply]
Hi! Sorry for the wait.

I did submit FSF papers a few years back. 

Should only checking visible text be the default? Or should it be configured as a setting or argument do you think?

It might be confusing if `jinx-correct-all` doesn't check all areas, but then again, you won't be able to see any corrections you are making unless jinx were to unfold it to show the text to the user.

**Reply by minad** [reply]
> Should only checking visible text be the default? Or should it be configured as a setting or argument do you think?

I'd say we only check the visible text and exclude folded text completely. We don't need a customization option.

This requires two major changes: (1) The commands should take visibility into account and only display visible misspellings for correction, (2) jinx--force-overlays calls jinx--check-region on the whole region, while it should rather only check visible areas like jinx--check-pending.

I expect that (2) leads to some more extensive refactoring of the existing checking infrastructure. I still hope that the code overall doesn't get much more complex since we already have visibility checking in jinx--check-pending. Please make sure that you don't introduce unnecessary code duplication.

> It might be confusing if jinx-correct-all doesn't check all areas, but then again, you won't be able to see any corrections you are making unless jinx were to unfold it to show the text to the user.

The current situation is not ideal, since we don't unfold hidden misspellings. The alternative would be to keep checking everything, and unfold when jumping to a hidden misspelling. This would also make the interaction convenient, but has the disadvantage that users cannot selectively check only visible text.

Checking everything is still easy - unfold everything at first, then start the check.

**Reply by minad** [reply]
Hmm, disregard what I wrote above. I got confused. I saw that we already perform unfolding!!

https://github.com/minad/jinx/blob/b2ef3af2857896e20cff0488c46a4bfcfee39ecc/jinx.el#L624-L645

This makes the premise of your suggestion invalid:

> It might be confusing if jinx-correct-all doesn't check all areas, but then again, you won't be able to see any corrections you are making unless jinx were to unfold it to show the text to the user.

Now I am less willing to follow through with your addition. Also note that I recently added region restriction in https://github.com/minad/jinx/commit/29a9e94ca3fe7068179384f2fc2929bcbc75b98c, so if you want to check only a certain region of text, this is already possible.

**Reply by walseb** [reply]
I see, so `jinx--check-region` should ideally be patched to only check visible areas. I will look into it. Making it work on a user level is simple enough though. 

I submitted a patch here with my previous changes https://github.com/minad/jinx/pull/150. I suppose that's better than nothing for now.

**Reply by minad** [reply]
> I suppose that's better than nothing for now.

No, sorry. The status quo is good enough right now. We check everything, but hidden areas are unfolded. See what I wrote in https://github.com/minad/jinx/discussions/144#discussioncomment-8735200. I closed https://github.com/minad/jinx/pull/150 for now. Please let's first finish the discussion here. I am not willing to consider trivial/half-baked patches.

**Reply by walseb** [reply]
> Hmm, disregard what I wrote above. I got confused. I saw that we already perform unfolding!!

I think that code is regarding highlighting. My issue only occurs when running a command like `jinx-correct-nearest`. It then tries to correct inside folds unless the patch I sent is applied. Is that correct or am I misunderstanding this?

**Reply by minad** [reply]
Jinx currently works like this:

1. `jinx-correct-nearest` only checks the visible area. I cannot reproduce your problem. Can you create a minimal recipe demonstrating the problem? (emacs -Q and only jinx-mode enabled)
2. `jinx-correct-all` checks everything (but respects marked regions) and **unfolds** hidden misspellings.

I believe this behavior is good enough and we don't have to change anything.

**Reply by walseb** [reply]
I think I understand the issue now. While `jinx-correct-all` unfolds org headings properly, it won't unfold all forms of hidden text. For instance, I use overlays with `(overlay-put new-overlay 'invisible t)` set, which `jinx` tries to check but doesn't seem to be able to unfold.

**Reply by minad** [reply]
I've found one bug though - `jinx-next` does not unfold the overlay! I will try to fix this.

EDIT: Fixed in https://github.com/minad/jinx/commit/bfeceb1441b8655b4723cf44937afe68d94740ce.

**Reply by minad** [reply]
> For instance, I use overlays with (overlay-put new-overlay 'invisible t) set, which jinx tries to check but doesn't seem to be able to unfold.

I see. This is a bug in your code. Your overlay should provide means for unfolding, e.g., via the Isearch unfolding API.

**Reply by walseb** [reply]
> I see. This is a bug in your code. Your overlay should provide means for unfolding, e.g., via the Isearch unfolding API.

Thanks, I didn't know that! Is there any way to indicate that the overlay shouldn't be unfolded?

**Reply by walseb** [reply]
I suppose I could add a face to it and put that face in `jinx-exclude-faces`.

**Reply by walseb** [reply]
The reason I don't want it unfolded and checked is because the overlay often contains many non-words which shouldn't be corrected.

**Reply by minad** [reply]
The correct approach seems then to exclude this region from checking in the first place via jinx-exclude-faces or a custom predicate function in jinx--predicates. You could for example attach a custom property jinx-exclude=t which is then be checked by a predicate function.

**Reply by walseb** [reply]
I see, thank you so much for the help! Attaching a face and putting it in `jinx-exclude-faces` seems good for now. I think this can be closed.

**Comment by walseb** [comment]
I think this should also apply to `jinx-next`

**Reply by minad** [reply]
Yes, all commands jumping to misspellings will be affected.
